File: sdk/javascript/src/events.ts
/**
 * Event Subscription System for Aura SDK
 *
 * Provides typed event subscriptions for blockchain events.
 */

import { WebSocket } from 'ws';

// ============================================================================
// Event Types
// ============================================================================

export interface BlockEvent {
  type: 'block';
  height: number;
  hash: string;
  timestamp: string;
  txCount: number;
}

export interface TxEvent {
  type: 'tx';
  hash: string;
  height: number;
  code: number;
  gasUsed: number;
  gasWanted: number;
  events: EventAttribute[];
}

export interface EventAttribute {
  type: string;
  attributes: { key: string; value: string }[];
}

// Module-specific events
export interface TransferEvent {
  type: 'transfer';
  sender: string;
  recipient: string;
  amount: string;
  denom: string;
}

export interface BridgeTransferEvent {
  type: 'bridge_transfer';
  transferId: string;
  sender: string;
  recipient: string;
  amount: string;
  targetChain: string;
  status: 'initiated' | 'completed' | 'failed';
}

export interface IdentityEvent {
  type: 'identity';
  action: 'created' | 'updated' | 'deleted';
  did: string;
  owner: string;
}

export interface GovernanceEvent {
  type: 'governance';
  action: 'proposal_submitted' | 'vote_cast' | 'proposal_passed' | 'proposal_rejected';
  proposalId: string;
  voter?: string;
}

export interface DEXEvent {
  type: 'dex';
  action: 'swap' | 'add_liquidity' | 'remove_liquidity' | 'order_created' | 'order_filled';
  poolId?: string;
  sender: string;
  amounts?: string[];
}

export type AuraEvent =
  | BlockEvent
  | TxEvent
  | TransferEvent
  | BridgeTransferEvent
  | IdentityEvent
  | GovernanceEvent
  | DEXEvent;

// ============================================================================
// Event Filter Types
// ============================================================================

export interface EventFilter {
  type?: string | string[];
  sender?: string;
  recipient?: string;
  module?: string;
  action?: string;
  minHeight?: number;
  maxHeight?: number;
}

export type EventHandler<T extends AuraEvent = AuraEvent> = (event: T) => void | Promise<void>;
export type ErrorHandler = (error: Error) => void;

// ============================================================================
// Event Subscriber
// ============================================================================

interface Subscription {
  id: string;
  filter: EventFilter;
  handler: EventHandler;
}

/**
 * EventSubscriber manages WebSocket connections and event subscriptions.
 */
export class EventSubscriber {
  private wsEndpoint: string;
  private ws: WebSocket | null = null;
  private subscriptions: Map<string, Subscription> = new Map();
  private errorHandler: ErrorHandler | null = null;
  private reconnectAttempts = 0;
  private maxReconnectAttempts = 5;
  private reconnectDelay = 1000;
  private isConnected = false;
  private subscriptionCounter = 0;

  constructor(wsEndpoint: string) {
    this.wsEndpoint = wsEndpoint;
  }

  /**
   * Connect to the WebSocket endpoint
   */
  async connect(): Promise<void> {
    return new Promise((resolve, reject) => {
      try {
        this.ws = new WebSocket(this.wsEndpoint);

        this.ws.onopen = () => {
          this.isConnected = true;
          this.reconnectAttempts = 0;
          this.resubscribeAll();
          resolve();
        };

        this.ws.onmessage = (event) => {
          this.handleMessage(event.data.toString());
        };

        this.ws.onerror = (error) => {
          if (this.errorHandler) {
            this.errorHandler(new Error(`WebSocket error: ${error.message}`));
          }
        };

        this.ws.onclose = () => {
          this.isConnected = false;
          this.attemptReconnect();
        };
      } catch (error) {
        reject(error);
      }
    });
  }

  /**
   * Disconnect from the WebSocket endpoint
   */
  disconnect(): void {
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
    this.isConnected = false;
    this.subscriptions.clear();
  }

  /**
   * Subscribe to events matching the filter
   */
  subscribe<T extends AuraEvent>(
    filter: EventFilter,
    handler: EventHandler<T>
  ): string {
    const id = `sub_${++this.subscriptionCounter}`;
    this.subscriptions.set(id, {
      id,
      filter,
      handler: handler as EventHandler,
    });

    if (this.isConnected && this.ws) {
      this.sendSubscription(filter);
    }

    return id;
  }

  /**
   * Unsubscribe from events
   */
  unsubscribe(subscriptionId: string): boolean {
    return this.subscriptions.delete(subscriptionId);
  }

  /**
   * Subscribe to new blocks
   */
  onBlock(handler: EventHandler<BlockEvent>): string {
    return this.subscribe({ type: 'block' }, handler);
  }

  /**
   * Subscribe to transactions
   */
  onTransaction(handler: EventHandler<TxEvent>, filter?: Partial<EventFilter>): string {
    return this.subscribe({ ...filter, type: 'tx' }, handler);
  }

  /**
   * Subscribe to transfers involving an address
   */
  onTransfer(
    address: string,
    handler: EventHandler<TransferEvent>
  ): string {
    return this.subscribe(
      { type: 'transfer', sender: address },
      handler
    );
  }

  /**
   * Subscribe to bridge transfer events
   */
  onBridgeTransfer(handler: EventHandler<BridgeTransferEvent>): string {
    return this.subscribe({ type: 'bridge_transfer', module: 'bridge' }, handler);
  }

  /**
   * Subscribe to identity events
   */
  onIdentity(handler: EventHandler<IdentityEvent>): string {
    return this.subscribe({ type: 'identity', module: 'identity' }, handler);
  }

  /**
   * Subscribe to governance events
   */
  onGovernance(handler: EventHandler<GovernanceEvent>): string {
    return this.subscribe({ type: 'governance', module: 'governance' }, handler);
  }

  /**
   * Subscribe to DEX events
   */
  onDEX(handler: EventHandler<DEXEvent>): string {
    return this.subscribe({ type: 'dex', module: 'dex' }, handler);
  }

  /**
   * Set error handler
   */
  onError(handler: ErrorHandler): void {
    this.errorHandler = handler;
  }

  /**
   * Check if connected
   */
  get connected(): boolean {
    return this.isConnected;
  }

  private handleMessage(data: string): void {
    try {
      const message = JSON.parse(data);
      const event = this.parseEvent(message);

      if (event) {
        for (const subscription of this.subscriptions.values()) {
          if (this.matchesFilter(event, subscription.filter)) {
            Promise.resolve(subscription.handler(event)).catch((err) => {
              if (this.errorHandler) {
                this.errorHandler(err);
              }
            });
          }
        }
      }
    } catch (error) {
      if (this.errorHandler) {
        this.errorHandler(error as Error);
      }
    }
  }

  private parseEvent(message: unknown): AuraEvent | null {
    if (!message || typeof message !== 'object') {
      return null;
    }

    const msg = message as Record<string, unknown>;
    const result = msg.result as Record<string, unknown> | undefined;

    if (!result || !result.data) {
      return null;
    }

    const data = result.data as Record<string, unknown>;
    const eventType = data.type as string;

    // Parse based on event type
    switch (eventType) {
      case 'tendermint/event/NewBlock':
        return this.parseBlockEvent(data);
      case 'tendermint/event/Tx':
        return this.parseTxEvent(data);
      default:
        return null;
    }
  }

  private parseBlockEvent(data: Record<string, unknown>): BlockEvent | null {
    const value = data.value as Record<string, unknown> | undefined;
    if (!value || !value.block) {
      return null;
    }

    const block = value.block as Record<string, unknown>;
    const header = block.header as Record<string, unknown>;

    return {
      type: 'block',
      height: parseInt(header.height as string, 10),
      hash: (block.last_commit as Record<string, unknown>)?.block_id as string || '',
      timestamp: header.time as string,
      txCount: ((block.data as Record<string, unknown>)?.txs as unknown[])?.length || 0,
    };
  }

  private parseTxEvent(data: Record<string, unknown>): TxEvent | null {
    const value = data.value as Record<string, unknown> | undefined;
    if (!value) {
      return null;
    }

    const txResult = value.TxResult as Record<string, unknown>;
    if (!txResult) {
      return null;
    }

    const result = txResult.result as Record<string, unknown>;

    return {
      type: 'tx',
      hash: txResult.hash as string || '',
      height: parseInt(txResult.height as string, 10),
      code: (result?.code as number) || 0,
      gasUsed: parseInt(result?.gas_used as string || '0', 10),
      gasWanted: parseInt(result?.gas_wanted as string || '0', 10),
      events: this.parseEventAttributes(result?.events as unknown[]),
    };
  }

  private parseEventAttributes(events: unknown[]): EventAttribute[] {
    if (!events || !Array.isArray(events)) {
      return [];
    }

    return events.map((e) => {
      const event = e as Record<string, unknown>;
      return {
        type: event.type as string,
        attributes: ((event.attributes as unknown[]) || []).map((a) => {
          const attr = a as Record<string, string>;
          return {
            key: attr.key || '',
            value: attr.value || '',
          };
        }),
      };
    });
  }

  private matchesFilter(event: AuraEvent, filter: EventFilter): boolean {
    if (filter.type) {
      const types = Array.isArray(filter.type) ? filter.type : [filter.type];
      if (!types.includes(event.type)) {
        return false;
      }
    }

    if (filter.minHeight && 'height' in event && event.height < filter.minHeight) {
      return false;
    }

    if (filter.maxHeight && 'height' in event && event.height > filter.maxHeight) {
      return false;
    }

    // Additional filter matching based on event type
    if (filter.sender && 'sender' in event && event.sender !== filter.sender) {
      return false;
    }

    if (filter.recipient && 'recipient' in event && event.recipient !== filter.recipient) {
      return false;
    }

    return true;
  }

  private sendSubscription(filter: EventFilter): void {
    if (!this.ws || !this.isConnected) {
      return;
    }

    // Build Tendermint query string
    const query = this.buildQuery(filter);
    const message = {
      jsonrpc: '2.0',
      method: 'subscribe',
      id: Date.now(),
      params: { query },
    };

    this.ws.send(JSON.stringify(message));
  }

  private buildQuery(filter: EventFilter): string {
    const conditions: string[] = [];

    if (filter.type === 'block') {
      conditions.push("tm.event='NewBlock'");
    } else if (filter.type === 'tx') {
      conditions.push("tm.event='Tx'");
    } else {
      conditions.push("tm.event='Tx'");
    }

    if (filter.module) {
      conditions.push(`message.module='${filter.module}'`);
    }

    if (filter.action) {
      conditions.push(`message.action='${filter.action}'`);
    }

    if (filter.sender) {
      conditions.push(`message.sender='${filter.sender}'`);
    }

    return conditions.join(' AND ');
  }

  private resubscribeAll(): void {
    for (const subscription of this.subscriptions.values()) {
      this.sendSubscription(subscription.filter);
    }
  }

  private attemptReconnect(): void {
    if (this.reconnectAttempts >= this.maxReconnectAttempts) {
      if (this.errorHandler) {
        this.errorHandler(new Error('Max reconnection attempts reached'));
      }
      return;
    }

    this.reconnectAttempts++;
    const delay = this.reconnectDelay * Math.pow(2, this.reconnectAttempts - 1);

    setTimeout(() => {
      this.connect().catch((error) => {
        if (this.errorHandler) {
          this.errorHandler(error);
        }
      });
    }, delay);
  }
}

// ============================================================================
// Factory Function
// ============================================================================

/**
 * Create an event subscriber for the given WebSocket endpoint
 */
export function createEventSubscriber(wsEndpoint: string): EventSubscriber {
  return new EventSubscriber(wsEndpoint);
}

File: sdk/python/aura/events.py
"""
Event Subscription System for Aura SDK

Provides typed event subscriptions for blockchain events.
"""

import asyncio
import json
from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Callable, Optional, Union
import websockets
from websockets.exceptions import ConnectionClosed


# ============================================================================
# Event Types
# ============================================================================


class EventType(str, Enum):
    """Enum of event types."""

    BLOCK = "block"
    TX = "tx"
    TRANSFER = "transfer"
    BRIDGE_TRANSFER = "bridge_transfer"
    IDENTITY = "identity"
    GOVERNANCE = "governance"
    DEX = "dex"


@dataclass
class BlockEvent:
    """New block event."""

    type: str = field(default="block")
    height: int = 0
    hash: str = ""
    timestamp: str = ""
    tx_count: int = 0


@dataclass
class TxEvent:
    """Transaction event."""

    type: str = field(default="tx")
    hash: str = ""
    height: int = 0
    code: int = 0
    gas_used: int = 0
    gas_wanted: int = 0
    events: list[dict[str, Any]] = field(default_factory=list)


@dataclass
class TransferEvent:
    """Token transfer event."""

    type: str = field(default="transfer")
    sender: str = ""
    recipient: str = ""
    amount: str = ""
    denom: str = ""


@dataclass
class BridgeTransferEvent:
    """Bridge transfer event."""

    type: str = field(default="bridge_transfer")
    transfer_id: str = ""
    sender: str = ""
    recipient: str = ""
    amount: str = ""
    target_chain: str = ""
    status: str = ""  # initiated, completed, failed


@dataclass
class IdentityEvent:
    """Identity event."""

    type: str = field(default="identity")
    action: str = ""  # created, updated, deleted
    did: str = ""
    owner: str = ""


@dataclass
class GovernanceEvent:
    """Governance event."""

    type: str = field(default="governance")
    action: str = ""  # proposal_submitted, vote_cast, proposal_passed, proposal_rejected
    proposal_id: str = ""
    voter: Optional[str] = None


@dataclass
class DEXEvent:
    """DEX event."""

    type: str = field(default="dex")
    action: str = ""  # swap, add_liquidity, remove_liquidity, order_created, order_filled
    pool_id: Optional[str] = None
    sender: str = ""
    amounts: list[str] = field(default_factory=list)


AuraEvent = Union[
    BlockEvent,
    TxEvent,
    TransferEvent,
    BridgeTransferEvent,
    IdentityEvent,
    GovernanceEvent,
    DEXEvent,
]


# ============================================================================
# Event Filter
# ============================================================================


@dataclass
class EventFilter:
    """Filter for event subscriptions."""

    type: Optional[Union[str, list[str]]] = None
    sender: Optional[str] = None
    recipient: Optional[str] = None
    module: Optional[str] = None
    action: Optional[str] = None
    min_height: Optional[int] = None
    max_height: Optional[int] = None


EventHandler = Callable[[AuraEvent], None]
AsyncEventHandler = Callable[[AuraEvent], Any]  # Coroutine
ErrorHandler = Callable[[Exception], None]


# ============================================================================
# Subscription
# ============================================================================


@dataclass
class Subscription:
    """Event subscription."""

    id: str
    filter: EventFilter
    handler: Union[EventHandler, AsyncEventHandler]


# ============================================================================
# Event Subscriber
# ============================================================================


class EventSubscriber:
    """Manages WebSocket connections and event subscriptions."""

    def __init__(self, ws_endpoint: str):
        self.ws_endpoint = ws_endpoint
        self._ws: Optional[websockets.WebSocketClientProtocol] = None
        self._subscriptions: dict[str, Subscription] = {}
        self._error_handler: Optional[ErrorHandler] = None
        self._reconnect_attempts = 0
        self._max_reconnect_attempts = 5
        self._reconnect_delay = 1.0
        self._is_connected = False
        self._subscription_counter = 0
        self._running = False

    async def connect(self) -> None:
        """Connect to the WebSocket endpoint."""
        try:
            self._ws = await websockets.connect(self.ws_endpoint)
            self._is_connected = True
            self._reconnect_attempts = 0
            await self._resubscribe_all()
        except Exception as e:
            if self._error_handler:
                self._error_handler(e)
            raise

    async def disconnect(self) -> None:
        """Disconnect from the WebSocket endpoint."""
        self._running = False
        if self._ws:
            await self._ws.close()
            self._ws = None
        self._is_connected = False
        self._subscriptions.clear()

    def subscribe(
        self,
        filter_: EventFilter,
        handler: Union[EventHandler, AsyncEventHandler],
    ) -> str:
        """Subscribe to events matching the filter."""
        self._subscription_counter += 1
        sub_id = f"sub_{self._subscription_counter}"

        self._subscriptions[sub_id] = Subscription(
            id=sub_id,
            filter=filter_,
            handler=handler,
        )

        return sub_id

    def unsubscribe(self, subscription_id: str) -> bool:
        """Unsubscribe from events."""
        if subscription_id in self._subscriptions:
            del self._subscriptions[subscription_id]
            return True
        return False

    def on_block(self, handler: Callable[[BlockEvent], None]) -> str:
        """Subscribe to new blocks."""
        return self.subscribe(EventFilter(type="block"), handler)

    def on_transaction(
        self,
        handler: Callable[[TxEvent], None],
        filter_: Optional[EventFilter] = None,
    ) -> str:
        """Subscribe to transactions."""
        f = filter_ or EventFilter()
        f.type = "tx"
        return self.subscribe(f, handler)

    def on_transfer(
        self,
        address: str,
        handler: Callable[[TransferEvent], None],
    ) -> str:
        """Subscribe to transfers involving an address."""
        return self.subscribe(
            EventFilter(type="transfer", sender=address),
            handler,
        )

    def on_bridge_transfer(
        self, handler: Callable[[BridgeTransferEvent], None]
    ) -> str:
        """Subscribe to bridge transfer events."""
        return self.subscribe(
            EventFilter(type="bridge_transfer", module="bridge"),
            handler,
        )

    def on_identity(self, handler: Callable[[IdentityEvent], None]) -> str:
        """Subscribe to identity events."""
        return self.subscribe(
            EventFilter(type="identity", module="identity"),
            handler,
        )

    def on_governance(self, handler: Callable[[GovernanceEvent], None]) -> str:
        """Subscribe to governance events."""
        return self.subscribe(
            EventFilter(type="governance", module="governance"),
            handler,
        )

    def on_dex(self, handler: Callable[[DEXEvent], None]) -> str:
        """Subscribe to DEX events."""
        return self.subscribe(
            EventFilter(type="dex", module="dex"),
            handler,
        )

    def on_error(self, handler: ErrorHandler) -> None:
        """Set error handler."""
        self._error_handler = handler

    @property
    def connected(self) -> bool:
        """Check if connected."""
        return self._is_connected

    async def run(self) -> None:
        """Run the event loop to process incoming messages."""
        self._running = True

        while self._running:
            try:
                if not self._ws or not self._is_connected:
                    await self.connect()

                async for message in self._ws:  # type: ignore
                    if not self._running:
                        break
                    await self._handle_message(str(message))

            except ConnectionClosed:
                self._is_connected = False
                if self._running:
                    await self._attempt_reconnect()
            except Exception as e:
                if self._error_handler:
                    self._error_handler(e)
                if self._running:
                    await self._attempt_reconnect()

    async def _handle_message(self, data: str) -> None:
        """Handle incoming WebSocket message."""
        try:
            message = json.loads(data)
            event = self._parse_event(message)

            if event:
                for subscription in self._subscriptions.values():
                    if self._matches_filter(event, subscription.filter):
                        try:
                            result = subscription.handler(event)
                            if asyncio.iscoroutine(result):
                                await result
                        except Exception as e:
                            if self._error_handler:
                                self._error_handler(e)
        except Exception as e:
            if self._error_handler:
                self._error_handler(e)

    def _parse_event(self, message: dict[str, Any]) -> Optional[AuraEvent]:
        """Parse WebSocket message into event."""
        result = message.get("result", {})
        data = result.get("data", {})

        if not data:
            return None

        event_type = data.get("type", "")

        if event_type == "tendermint/event/NewBlock":
            return self._parse_block_event(data)
        elif event_type == "tendermint/event/Tx":
            return self._parse_tx_event(data)

        return None

    def _parse_block_event(self, data: dict[str, Any]) -> Optional[BlockEvent]:
        """Parse block event."""
        value = data.get("value", {})
        block = value.get("block", {})
        header = block.get("header", {})

        if not header:
            return None

        return BlockEvent(
            height=int(header.get("height", 0)),
            hash=block.get("last_commit", {}).get("block_id", ""),
            timestamp=header.get("time", ""),
            tx_count=len(block.get("data", {}).get("txs", [])),
        )

    def _parse_tx_event(self, data: dict[str, Any]) -> Optional[TxEvent]:
        """Parse transaction event."""
        value = data.get("value", {})
        tx_result = value.get("TxResult", {})

        if not tx_result:
            return None

        result = tx_result.get("result", {})

        return TxEvent(
            hash=tx_result.get("hash", ""),
            height=int(tx_result.get("height", 0)),
            code=result.get("code", 0),
            gas_used=int(result.get("gas_used", 0)),
            gas_wanted=int(result.get("gas_wanted", 0)),
            events=result.get("events", []),
        )

    def _matches_filter(self, event: AuraEvent, filter_: EventFilter) -> bool:
        """Check if event matches filter."""
        if filter_.type:
            types = [filter_.type] if isinstance(filter_.type, str) else filter_.type
            if event.type not in types:
                return False

        if filter_.min_height and hasattr(event, "height"):
            if event.height < filter_.min_height:  # type: ignore
                return False

        if filter_.max_height and hasattr(event, "height"):
            if event.height > filter_.max_height:  # type: ignore
                return False

        if filter_.sender and hasattr(event, "sender"):
            if event.sender != filter_.sender:  # type: ignore
                return False

        if filter_.recipient and hasattr(event, "recipient"):
            if event.recipient != filter_.recipient:  # type: ignore
                return False

        return True

    async def _send_subscription(self, filter_: EventFilter) -> None:
        """Send subscription request."""
        if not self._ws or not self._is_connected:
            return

        query = self._build_query(filter_)
        message = {
            "jsonrpc": "2.0",
            "method": "subscribe",
            "id": int(asyncio.get_event_loop().time() * 1000),
            "params": {"query": query},
        }

        await self._ws.send(json.dumps(message))

    def _build_query(self, filter_: EventFilter) -> str:
        """Build Tendermint query string."""
        conditions: list[str] = []

        if filter_.type == "block":
            conditions.append("tm.event='NewBlock'")
        elif filter_.type == "tx":
            conditions.append("tm.event='Tx'")
        else:
            conditions.append("tm.event='Tx'")

        if filter_.module:
            conditions.append(f"message.module='{filter_.module}'")

        if filter_.action:
            conditions.append(f"message.action='{filter_.action}'")

        if filter_.sender:
            conditions.append(f"message.sender='{filter_.sender}'")

        return " AND ".join(conditions)

    async def _resubscribe_all(self) -> None:
        """Resubscribe to all subscriptions after reconnection."""
        for subscription in self._subscriptions.values():
            await self._send_subscription(subscription.filter)

    async def _attempt_reconnect(self) -> None:
        """Attempt to reconnect with exponential backoff."""
        if self._reconnect_attempts >= self._max_reconnect_attempts:
            if self._error_handler:
                self._error_handler(Exception("Max reconnection attempts reached"))
            self._running = False
            return

        self._reconnect_attempts += 1
        delay = self._reconnect_delay * (2 ** (self._reconnect_attempts - 1))

        await asyncio.sleep(delay)

        try:
            await self.connect()
        except Exception as e:
            if self._error_handler:
                self._error_handler(e)


def create_event_subscriber(ws_endpoint: str) -> EventSubscriber:
    """Create an event subscriber for the given WebSocket endpoint."""
    return EventSubscriber(ws_endpoint)
